Specification fixes and rewriting (#181)
eclipse-biscuit/biscuit: SPECIFICATIONS.md

  repeated SignedBlock blocks = 3;
  required Proof proof = 4;
}
```

> The `rootKeyId` is a hint to decide which root public key should be used

Reviewer: ```suggestion
The `rootKeyId` is a hint to help the consumer decide which root public key should be used
```

for signature verification.
The `authority` block has a special meaning, as it is signed by the root key,
so it is stored separately.
The `blocks` field contains the rest of the block list, that must stay in the same order.
The `proof` field contains either:
- `nextSecret`: the private key corresponding to the `nextKey` public key in the last block (attenuable tokens)
- `finalSignature`: a signature of the last block by that private key (sealed tokens).

```proto
message Proof {
  oneof Content {
    bytes nextSecret = 1;
    bytes finalSignature = 2;
  }
}
```

> The `SignedBlock` structure carries the sigend data and its signatures, representing the

Reviewer: ```suggestion
The `SignedBlock` structure carries the signed data and its signatures, representing the
```

chain of trust starting from the root key.

```proto
message SignedBlock {
  required bytes block = 1;
  required PublicKey nextKey = 2;
  required bytes signature = 3;
  optional ExternalSignature externalSignature = 4;
  optional uint32 version = 5;
}
```

Each `SignedBlock` contains:

> - `block`: a `Block` structure serialized as a byte array

Reviewer: maybe we should hint that why there are two layers of serialization (unless it’s already done somewhere else)

- `nextKey`: the public key used to verify the next `SignedBlock`
- `signature`: the signature of the current block
- `externalSignature`: if present, it is a signature of the current block by another key
- `version`: indicates the version of the signed payload format

> The version field is used to vary how the block is signed and verified.

Reviewer: ```suggestion
The version field is used to specify how the block was signed and must be verified.
```

  required bytes key = 2;
}
```

Public keys are serialized as a byte array, with the `algorithm` field discriminating between

> algorithms. The algorithm depedent serialization format is described in the [Cryptography section](#Algorithms).

Reviewer: ```suggestion
algorithms. The algorithm-dependent serialization format is described in the [Cryptography section](#Algorithms).
```

}
```

Each `Block` contains the following fields:
- `symbols`: the list of new [symbols](#symbol-table) introduced in this block.

> - `context`: a free form field with no particular meaning for Biscuit authorization. It can be used to hold application specific data

Reviewer: ```suggestion
- `context`: a free-form field with no particular meaning for Biscuit authorization. It can be used to hold application specific data
```

### Text format

When transmitted as text, a Biscuit token should be serialized to a
URL safe base 64 string. When the context does not indicate that it
is a Biscuit token, that base 64 string should be prefixed with `biscuit:`.

> ## Cryptography

Author: Putting the cryptography section at the same level as the format section

- The lowest supported datalog version is `v3.0`;
- The highest supported datalog version is `v3.3`;

### Text format

When transmitted as text, a Biscuit token should be serialized to a

> URL safe base 64 string. When the context does not indicate that it

Reviewer: ```suggestion
URL-safe base 64 string. When the context does not indicate that it
```

If all signatures are verified, extract pk_n+1 from the last block and
sig from the proof field, and check `verify(pk_n+1, sig_n+1, data_n+alg_n+1+pk_n+1+sig_n)`

> ### Blocks

Author: this is already covered in a different part of the spec

> ## Symbol table

Author: the symbol table section is moving up a level. Mayb it should move under the format section though?

Reviewer: maybe we should rename it to "string interning", since it’s not just a serialization matter
